Mathlib/Order/OrderType.lean

```diff
@@ -0,0 +1,450 @@
+
+@[simp]
+theorem card_one : card 1 = 1 := mk_eq_one _
+
+end Cardinal
+/-- `ω` is the first infinite ordinal, defined as the order type of `ℕ`. -/
+def omega0 : OrderType := type ℕ
+
+/-- The order type of the rational numbers. -/
+def eta : OrderType := type ℚ
+
+/-- The order type of the real numbers on the interval `(0,1)`. -/
+def theta : OrderType := type (Set.Ioo (0 : ℝ) 1)
+
+@[inherit_doc]
+scoped notation "ω" => OrderType.omega0
+
+@[inherit_doc]
+scoped notation "η" => OrderType.eta
+
+@[inherit_doc]
+scoped notation "θ" => OrderType.theta
+
+open Classical
+in instance : Add OrderType.{u} where
+  add := Quotient.map₂ (fun r s ↦ ⟨(r ⊕ₗ s)⟩)
+   (fun _ _ ha _ _ hb ↦ ⟨OrderIso.sumLexCongr (choice ha) (choice hb)⟩)
+
+open Classical
+in instance : HAdd OrderType.{u} OrderType.{v} OrderType.{max u v} where
+  hAdd := Quotient.map₂ (fun r s ↦ ⟨(r ⊕ₗ s)⟩)
+   (fun _ _ ha _ _ hb ↦ ⟨OrderIso.sumLexCongr (choice ha) (choice hb)⟩)
+
+@[simp]
+lemma type_add (α : Type u) (β : Type v) [LinearOrder α] [LinearOrder β] :
+    type (α ⊕ₗ β) = (type α) + (type β) := rfl
+
+lemma OrderIso.sumLexEmpty (α : Type u) [LinearOrder α] : Nonempty (Lex (α ⊕ PEmpty) ≃o α) :=
+  ⟨OrderIso.ofRelIsoLT ((Sum.Lex.toLexRelIsoLT (α := α) (β := PEmpty)).symm.trans
+    (RelIso.sumLexEmpty (β := PEmpty) (α := α) (r := (· < ·)) (s := (· < ·))))⟩
+
+lemma OrderIso.emptySumLex (α : Type u) [LinearOrder α] : Nonempty (Lex (PEmpty ⊕ α) ≃o α) :=
+   ⟨OrderIso.ofRelIsoLT ((Sum.Lex.toLexRelIsoLT (α := PEmpty) (β := α)).trans
+     (RelIso.emptySumLex (β := α) (α := PEmpty) (r := (· < ·)) (s := (· < ·))))⟩
+
+open Classical in
+lemma add_zero (o : OrderType.{u}) : o + 0 = o :=
+  inductionOn o (fun α _ ↦ RelIso.ordertype_congr (choice (OrderIso.sumLexEmpty α)))
+
+open Classical in
+lemma zero_add (o : OrderType.{u}) : 0 + o = o :=
+  inductionOn o (fun α _ ↦ RelIso.ordertype_congr (choice (OrderIso.emptySumLex α)))
+
+open Classical in
+lemma add_assoc (o₁ o₂ o₃ : OrderType.{u}) : o₁ + o₂ + o₃ = o₁ + (o₂ + o₃) :=
+  inductionOn₃ o₁ o₂ o₃ (fun α _ β _ γ _ ↦ RelIso.ordertype_congr (OrderIso.sumLexAssoc α β γ))
+
+instance : AddMonoid OrderType where
+  add_assoc := add_assoc
+  zero_add := zero_add
+  add_zero := add_zero
+  nsmul := nsmulRec
+
+/-- `Equiv.prodCongr` promoted to an order isomorphism between lexicographic products. -/
+@[simps! apply]
+def OrderIso.prodLexCongr [LinearOrder α] [LinearOrder β]
+    [LinearOrder γ] [LinearOrder δ] (ea : α ≃o β) (eb : γ ≃o δ) : α ×ₗ γ ≃o β ×ₗ δ where
+  toEquiv := ofLex.trans ((Equiv.prodCongr ea eb).trans toLex)
+  map_rel_iff' := by
+    intro a b
+    simp [Prod.Lex.le_iff, OrderIso.lt_iff_lt]
+
+open Classical in
+instance : Mul OrderType where
+  mul := Quotient.map₂ (fun r s ↦ ⟨(r ×ₗ s)⟩)
+   (fun _ _ ha _ _ hb ↦ ⟨OrderIso.prodLexCongr (choice ha) (choice hb)⟩)
+
+open Classical in
+instance : HMul OrderType.{u} OrderType.{v} OrderType.{max u v} where
+  hMul := Quotient.map₂ (fun r s ↦ ⟨(r ×ₗ s)⟩)
```

Comment:
This needs to be the other way around! Unfortunately Cantor really wanted his ordinal functions to all be normal on the right argument, so now we're stuck writing `ω * 2` until the end of time.
```suggestion
  hMul := Quotient.map₂ (fun r s ↦ ⟨(s ×ₗ r)⟩)
```